### PR TITLE
ddata ReadMajorityPlus and WriteMajorityPlus

### PR DESCRIPTION
**Pull request by zbynek001**
New ReadMajorityPlus and WriteMajorityPlus
Possibility to prefer oldest in ddata writes and reads 
(partially migrated from https://github.com/akka/akka/pull/28895)

### CODE REVIEW

File: src/contrib/cluster/Akka.DistributedData.Tests/ReplicatorResiliencySpec.cs [hunk 1]
```diff
@@ -180,15 +180,16 @@ public async Task DistributedData_Replicator_Defaults_to_NoSupervisor()
             Watch(durableStore);
             durableStore.Tell(new InitFail());
 
-            ExpectTerminated(durableStore, TimeSpan.FromSeconds(10));
-            ExpectTerminated(replicator, TimeSpan.FromSeconds(10));
+            var terminated1 = ExpectMsg<Terminated>(TimeSpan.FromSeconds(10));
+            var terminated2 = ExpectMsg<Terminated>(TimeSpan.FromSeconds(10));
+            ImmutableHashSet.Create(terminated1.ActorRef, terminated2.ActorRef).Should().BeEquivalentTo(durableStore, replicator);
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.DistributedData.Tests/ReplicatorResiliencySpec.cs:185]: Nice, that should make this test less racy

File: src/contrib/cluster/Akka.DistributedData.Tests/WriteAggregatorSpec.cs [hunk 2]
```diff
@@ -85,7 +85,7 @@ public WriteAckAdapter(IActorRef replica)
         private readonly Address _nodeB = new Address("akka.tcp", "Sys", "b", 2552);
         private readonly Address _nodeC = new Address("akka.tcp", "Sys", "c", 2552);
         private readonly Address _nodeD = new Address("akka.tcp", "Sys", "d", 2552);
-        private readonly IImmutableSet<Address> _nodes;
+        private readonly IImmutableList<Address> _nodes;
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.DistributedData.Tests/WriteAggregatorSpec.cs:88]: Why the change from set to list?

File: src/contrib/cluster/Akka.DistributedData/ReadAggregator.cs [hunk 3]
```diff
@@ -41,16 +41,29 @@ public ReadAggregator(IKey key, IReadConsistency consistency, object req, IImmut
...
+                        Log.Debug("ReadMajority [{0}] [{1}] of [{2}].", _key, r, n);
+                        return n - r;
+                    }
+                case ReadMajorityPlus read:
+                    {
+                        // +1 because local node is not included in 'Nodes'
+                        var n = Nodes.Count + 1;
+                        var r = CalculateMajority(read.MinCapacity, n, read.Additional);
+                        Log.Debug("ReadMajorityPlus [{0}] [{1}] of [{2}].", _key, r, n);
+                        return n - r;
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.DistributedData/ReadAggregator.cs:62]: LGTM

File: src/contrib/cluster/Akka.DistributedData/ReadWriteAggregator.cs [hunk 4]
```diff
@@ -38,46 +40,53 @@ protected sealed class SendToSecondary
...
                 {
-                    var n = Nodes.OrderBy(x => ThreadLocalRandom.Current.Next()).ToArray();
-                    var p = n.Take(primarySize).ToImmutableHashSet();
-                    var s = n.Skip(primarySize).Take(MaxSecondaryNodes).ToImmutableHashSet();
-                    return ((IImmutableSet<Address>)p, (IImmutableSet<Address>)s);
+                    // Prefer to use reachable nodes over the unreachable nodes first.
+                    // When RequiresCausalDeliveryOfDeltas (shuffle=false) use deterministic order to so that sequence numbers
+                    // of subsequent updates are in sync on the destination nodes.
+                    // The order is also kept when prefer-oldest is enabled.
+                    var orderedNodes = Shuffle ?
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.DistributedData/ReadWriteAggregator.cs:73]: The comment above makes sense - thank you

File: src/contrib/cluster/Akka.DistributedData/ReadWriteAggregator.cs [hunk 5]
```diff
@@ -38,46 +40,53 @@ protected sealed class SendToSecondary
...
         }
 
-        public static int CalculateMajorityWithMinCapacity(int minCapacity, int numberOfNodes)
+        public static int CalculateMajority(int minCapacity, int numberOfNodes, int additional)
         {
-            if (numberOfNodes <= minCapacity) return numberOfNodes;
-            
-            return Math.Max(minCapacity, numberOfNodes / 2 + 1);
+            var majority = numberOfNodes / 2 + 1;
+            return Math.Min(numberOfNodes, Math.Max(majority + additional, minCapacity));
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.DistributedData/ReadWriteAggregator.cs:88]: LGTM

File: src/contrib/cluster/Akka.DistributedData/Replicator.cs [hunk 6]
```diff
@@ -289,19 +289,23 @@ internal sealed class Replicator : UntypedActor, IWithUnboundedStash
...
+        private ImmutableSortedSet<Member> _membersByAge = ImmutableSortedSet<Member>.Empty.WithComparer(Member.AgeOrdering);
 
-        private ImmutableHashSet<Address> AllNodes => _nodes.Union(_weaklyUpNodes);
+        private ImmutableSortedSet<Address> AllNodes => _nodes.Union(_weaklyUpNodes);
 
         /// <summary>
         /// Cluster weaklyUp nodes, doesn't contain joining and not selfAddress
         /// </summary>
-        private ImmutableHashSet<Address> _weaklyUpNodes = ImmutableHashSet<Address>.Empty;
+        private ImmutableSortedSet<Address> _weaklyUpNodes = ImmutableSortedSet<Address>.Empty;
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.DistributedData/Replicator.cs:303]: Moving to sorted set - part of the deterministic sort order mentioned earlier. Got it.

File: src/contrib/cluster/Akka.DistributedData/Replicator.cs [hunk 7]
```diff
@@ -419,6 +423,14 @@ public Replicator(ReplicatorSettings settings)
             else Become(NormalReceive);
         }
 
+        private IImmutableList<Address> NodesForReadWrite()
+        {
+            if (_settings.PreferOldest)
+                return _membersByAge.Select(i => i.Address).ToImmutableList();
+            else
+                return _nodes.ToImmutableList();
+        }
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.DistributedData/Replicator.cs:432]: How frequently is this method called, out of curiosity?

File: src/contrib/cluster/Akka.DistributedData/Replicator.cs [hunk 8]
```diff
@@ -601,7 +613,7 @@ private void ReceiveGet(IKey key, IReadConsistency consistency, object req)
                 else Sender.Tell(new GetSuccess(key, req, localValue.Data));
             }
             else
-                Context.ActorOf(ReadAggregator.Props(key, consistency, req, _nodes, _unreachable, localValue, Sender)
+                Context.ActorOf(ReadAggregator.Props(key, consistency, req, NodesForReadWrite(), _unreachable, !_settings.PreferOldest, localValue, Sender)
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.DistributedData/Replicator.cs:616]: So on each read aggregation we might run that query - ok, don't think that's all that frequent for the most part

Comment (Aaronontheweb) [reply src/contrib/cluster/Akka.DistributedData/Replicator.cs:616]: Shuffle is enabled when `PreferOldest == false` ?

File: src/contrib/cluster/Akka.DistributedData/Replicator.cs [hunk 9]
```diff
@@ -1090,14 +1108,12 @@ private void GossipTo(Address address)
...
         }
 
-        private Address SelectRandomNode(ImmutableHashSet<Address> addresses)
+        private Address SelectRandomNode(IReadOnlyList<Address> addresses)
         {
             if (addresses.Count > 0)
             {
-                var random = ThreadLocalRandom.Current.Next(addresses.Count - 1);
-                return addresses.Skip(random).First();
+                return addresses[ThreadLocalRandom.Current.Next(addresses.Count)];
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.DistributedData/Replicator.cs:1115]: So is this why we're using a list rather than a hash set? To be able to use the indexer?

File: src/contrib/cluster/Akka.DistributedData/ReplicatorSettings.cs [hunk 10]
```diff
@@ -78,11 +78,12 @@ public static ReplicatorSettings Create(Config config)
                 pruningMarkerTimeToLive: config.GetTimeSpan("pruning-marker-time-to-live", TimeSpan.FromHours(6)),
                 durablePruningMarkerTimeToLive: durableConfig.GetTimeSpan("pruning-marker-time-to-live", TimeSpan.FromDays(10)),
                 maxDeltaSize: config.GetInt("delta-crdt.max-delta-size", 50),
-                restartReplicatorOnFailure: config.GetBoolean("recreate-on-failure", false));
+                restartReplicatorOnFailure: config.GetBoolean("recreate-on-failure", false),
+                preferOldest: config.GetBoolean("prefer-oldest"));
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.DistributedData/ReplicatorSettings.cs:82]: Got it

File: src/contrib/cluster/Akka.DistributedData/Replicator.cs [hunk 11]
```diff
@@ -1498,11 +1518,11 @@ protected override ImmutableArray<Address> AllNodes
             {
                 get
                 {
-                    var allNodes = _replicator.AllNodes.Except(_replicator._unreachable).OrderBy(x => x).ToImmutableArray();
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.DistributedData/Replicator.cs:1501]: No longer need to sort because underlying set is sorted. Got it.

File: src/contrib/cluster/Akka.DistributedData/Serialization/ReplicatorMessageSerializer.cs [hunk 12]
```diff
@@ -385,7 +385,24 @@ private Proto.Msg.Get GetToProto(Get msg)
...
+                    if (rmp.MinCapacity != 0)
+                    {
+                        proto.ConsistencyMinCap = rmp.MinCapacity;
+                        proto.HasConsistencyMinCap = true;
+                    }
+                    if(rmp.Additional != 0)
+                    {
+                        proto.ConsistencyAdditional = rmp.Additional;
+                        proto.HasConsistencyAdditional = true;
+                    }
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.DistributedData/Serialization/ReplicatorMessageSerializer.cs:405]: LGTM

File: src/contrib/cluster/Akka.DistributedData/WriteAggregator.cs [hunk 13]
```diff
@@ -65,12 +65,24 @@ private int GetDoneWhenRemainingSize()
...
+                    }
+                case WriteMajorityPlus write:
+                    {
+                        // +1 because local node is not included in 'Nodes'
+                        var n = Nodes.Count + 1;
+                        var w = CalculateMajority(write.MinCapacity, n, write.Additional);
+                        Log.Debug("WriteMajorityPlus [{0}] [{1}] of [{2}].", _key, w, n);
+                        return n - w;
+                    }
+                case WriteLocal _: throw new ArgumentException("WriteAggregator does not support WriteLocal");
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.DistributedData/WriteAggregator.cs:85]: LGTM

File: src/contrib/cluster/Akka.DistributedData/WriteAggregator.cs [hunk 14]
```diff
@@ -252,6 +265,48 @@ public override int GetHashCode()
         }
     }
 
+    /// <summary>
+    /// <see cref="WriteMajority"/> but with the given number of <see cref="Additional"/> nodes added to the majority count. At most
+    /// all nodes.
+    /// </summary>
+    public sealed class WriteMajorityPlus : IWriteConsistency, IEquatable<WriteMajorityPlus>
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.DistributedData/WriteAggregator.cs:272]: LGTM

File: src/contrib/cluster/Akka.DistributedData/reference.conf [hunk 15]
```diff
@@ -25,78 +25,82 @@ akka.cluster.distributed-data {
...
-  # When set, this flag will attach a backoff supervisor to the replicator; 
+
+  # When set, this flag will attach a backoff supervisor to the replicator;
   # any failing replicator to be restarted
   recreate-on-failure = off
-  
+
+  # Update and Get operations are sent to oldest nodes first.
+  # This is useful together with Cluster Singleton, which is running on oldest nodes.
+  prefer-oldest = off
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.DistributedData/reference.conf:63]: LGTM

File: src/protobuf/ReplicatorMessages.proto [hunk 16]
```diff
@@ -17,6 +17,7 @@ message Get {
    int32 consistencyMinCap = 5;
    bool hasConsistencyAdditional = 6;
    int32 consistencyAdditional = 7;
+   bool hasConsistencyMinCap = 8;
```
Comment (Aaronontheweb) [inline src/protobuf/ReplicatorMessages.proto:20]: LGTM

**Review by Aaronontheweb — COMMENTED** [review]
Overall, looks good to me - left you some questions though @zbynek001

File: src/contrib/cluster/Akka.DistributedData/Replicator.cs (hunk 8 above)
Comment (zbynek001) [reply src/contrib/cluster/Akka.DistributedData/Replicator.cs:616]: Exactly. When `PreferOldest == true`, nodes are sorted by oldest.
When `PreferOldest == false`, nodes are taken in a random order

File: src/contrib/cluster/Akka.DistributedData/Replicator.cs (hunk 9 above)
Comment (zbynek001) [reply src/contrib/cluster/Akka.DistributedData/Replicator.cs:1115]: The real input type of `addresses` is `ImmutableSortedSet<>`. The `IReadOnlyList` in this method is to be able to use the index

File: src/contrib/cluster/Akka.DistributedData.Tests/WriteAggregatorSpec.cs (hunk 2 above)
Comment (zbynek001) [reply src/contrib/cluster/Akka.DistributedData.Tests/WriteAggregatorSpec.cs:88]: because the base class `ReadWriteAggregator` changes of `nodes` from `IImmutableSet` to `IImmutableList` to keep the order of nodes by oldest for the `PreferOldest` scenario.

File: src/contrib/cluster/Akka.DistributedData/Replicator.cs (hunk 7 above)
Comment (zbynek001) [reply src/contrib/cluster/Akka.DistributedData/Replicator.cs:432]: if you're worried about the `ToImmutableList` being called to often, another option for `PreferOldest==false` scenario would be to switch to `IReadOnlyList` instead of `IImmutableList`. Then we could skip it

File: src/contrib/cluster/Akka.DistributedData.Tests/WriteAggregatorSpec.cs (hunk 2 above)
Comment (Aaronontheweb) [reply src/contrib/cluster/Akka.DistributedData.Tests/WriteAggregatorSpec.cs:88]: Got it, thank you

**Review by Aaronontheweb — APPROVED** [review]
LGTM

File: src/contrib/cluster/Akka.DistributedData/Replicator.cs (hunk 7 above)
Comment (Aaronontheweb) [reply src/contrib/cluster/Akka.DistributedData/Replicator.cs:432]: We can optimize it in a subsequent PR if it's an issue
